boilerplate/_lib/boilerplate-commit

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env bash
+
+set -e
+
+REPO_ROOT=$(git rev-parse --show-toplevel)
+source $REPO_ROOT/boilerplate/_lib/common.sh
+
+tmpd=$(mktemp -d)
+trap "rm -fr $tmpd" EXIT
+git_status=$tmpd/git_status
+bp_log=$tmpd/bp_git_log
+bp_clone=$tmpd/bp_git_clone
+convention_status=$tmpd/convention_status
+commit_message=$tmpd/commit_msg
+
+# Variables to keep track of what's happening in this commit. Empty
+# means we're not doing whatever it is.
+####
+# - Bootstrapping: bringing boilerplate into the repo for the first
+# time. Nonempty if bootstrapping.
+bootstrap=
+####
+# - If we were already bootstrapped, and boilerplate-update brought in a
+# newer boilerplate commit, we'll put "{from_hash}...{to_hash}" here.
+# This should be mutually exclusive with `bootstrap`.
+bp_commit_change=
+####
+# - Changes in conventions. This is a file containing one line per
+# convention indicating what was done in this commit with respect to
+# that convention: "Subscribe", "Update", or "No change". (TODO:
+# "Unsubscribe".) The file is only empty if update.cfg is
+# (substantively) empty.
+convention_statuses=$tmpd/convention_statuses
+>$convention_statuses
+####
+
+git status --porcelain > $git_status
+
+# Bootstrapping includes adding the boilerplate-update target to the
+# Makefile and adding boilerplate/update and boilerplate/update.cfg. We
+# won't bother with the former. Since the latter are new files in a new
+# directory, `git status` will just show the `boilerplate/` directory as
+# untracked.
+if grep -q '^?? boilerplate/$' $git_status; then
+  bootstrap=true
+
+# This wasn't a bootstrap. We can detect it was an update if the
+# last-boilerplate-commit file was changed.
+elif grep -q '^ M boilerplate/_data/last-boilerplate-commit$' $git_status; then
+  # Produce a string of the form {old_hash}...{new_hash}
+  bp_commit_change=$(git diff boilerplate/_data/last-boilerplate-commit | tail -2 | paste -d/ -s - | sed 's/[+-]//g; s,/,...,')
+  # Handy URL showing the commits and deltas
+  bp_compare_url="https://github.com/openshift/boilerplate/compare/$bp_commit_change"
+  # Generate the commit history for this range. This will go in the commit message.
+  (
+    if [ -z "$BOILERPLATE_GIT_REPO" ]; then
```

Comment:
Wouldn't it be good to factorize this behaviour (as already present in `freeze-check` and `update` ) ?

Reply:
Well, I would love to factor it out of `update`, but as explained [below](https://github.com/openshift/boilerplate/pull/56/files/9c5adcc6a9026422324e77d0fbf8c4c3dc2df90f#diff-2f0d8358cb701c26357e5dc568144e13R73-R76), that needs to be self-contained as it's part of the bootstrap process.

I didn't recall that this logic was in `freeze-check`. Yes, it would be good to pull this into a lib function common to both.

Reply:
Done.